### PR TITLE
Load maps from server

### PR DESCRIPTION
**Pull request by eric79**
This allows the app to replace the hard-coded pre-loaded maps with those that users have uploaded.  There is a corresponding branch with the same name on the app side.

@davideaster would you mind testing and reviewing both?

@AmbientOSX @landersk61 @youngca @longtinm , fyi

### CODE REVIEW

File: lib/nodejs/lobby.js [hunk 1]
```diff
@@ -44,6 +44,24 @@ app.get( "/", function( req, res ) {
   } );
 } );
 
+app.get( "/map-list", function( req, res ) {
+  var mapDir = "documents/ITDG/maps";
+
+  // Ensure that the map directory exists, then get the list of files in the map directory
+  mkdirp( mapDir ).then( function( err ) {
```
Comment (davideaster) [inline lib/nodejs/lobby.js]: The promise doesn't pass an error to the callback (well, [mostly](http://exploringjs.com/es6/ch_promises.html#_mistake-using-then-for-error-handling)). Errors will pass through the chain and can be handled in a `catch` clause at the end.

File: lib/nodejs/lobby.js [hunk 2]
```diff
@@ -44,6 +44,24 @@ app.get( "/", function( req, res ) {
...
+    if ( err ) {
+      console.error( err );
+      res.status( 500 ).send( err );
```
Comment (davideaster) [inline lib/nodejs/lobby.js]: In Express, unless the request has a specific response to an error, it's better to pass the error on to `next`. Express will stop routing at that point and will show the application's error page.  Development-mode servers will show the stack trace, and production-mode servers won't, for example.

File: lib/nodejs/lobby.js (hunk 1 above)
Comment (davideaster) [inline lib/nodejs/lobby.js]: A little more complicated and probably not worth the change, but you could skip `mkdirp` and go straight to `readdirAsync`. You would need to catch `ENOENT` after `readdirAsync` and return an empty result. It's a little unexpected for a GET to change the server.

File: lib/nodejs/lobby.js [hunk 3]
```diff
@@ -44,6 +44,24 @@ app.get( "/", function( req, res ) {
...
+      return;
+    }
+    fs.readdirAsync( mapDir ).then( function( files ) {
+      res.send( {
+        mapFilenames: files || []
+      } );
+    } );
+  } );
+} );
```
Comment (davideaster) [inline lib/nodejs/lobby.js:55]: Here's an example with a `catch` handler at the end and forwarding errors to `done`:

```
app.get( "/map-list", function( req, res, next ) {
  var mapDir =
    "documents/ITDG/maps";
  mkdirp( mapDir ).
    then( made => fs.readdirAsync( mapDir ) ).
    then( files => res.send( { mapFilenames: files || [] } ) ).
    catch( err => next( err ) );
} );
```

**Comment by eric79** [comment]
@davideaster you are going to drag me kicking and screaming into this new ES6 syntax.  :wink: Thanks for the review and the suggestions.  It appears to be working well.  :smile:

**Pull request merged**